src/pool/poolClient.ts

```diff
@@ -362,6 +414,46 @@ export class Client {
   private getOrCreatePoolEvents() {
     return this.poolEvents
   }
+  private getOrCreateUserService(userAddress:string,tokenAddress:string){
+    if(has(this.userServices,[tokenAddress,userAddress])) return get(this.userServices,[tokenAddress,userAddress]);
+    const erc20Contract = this.getOrCreateErc20Contract(tokenAddress)
+    const userService = new UserState(erc20Contract);
+    // this service is stateful now, so needs to be cached
+    set(this.userServices,[tokenAddress,userAddress],userService)
+    return userService
+  }
+  private updateExchangeRateTable(l1TokenAddress:string, exchangeRateTable:Record<number,BigNumberish>):Record<number,BigNumberish>{
+    if(!this.exchangeRateTable[l1TokenAddress]) this.exchangeRateTable[l1TokenAddress] = {}
+    this.exchangeRateTable[l1TokenAddress] = {...this.exchangeRateTable[l1TokenAddress], ...exchangeRateTable}
+    return this.exchangeRateTable[l1TokenAddress]
+  }
+  // calculates the value of each LP token transfer at the block it was sent. this only works if we have archive node
+  async calculateLpTransferValue(l1TokenAddress:string, userState: Awaited<ReturnType<UserState["read"]>>){
```

Comment:
this is the main part which allows us to calcluate fees. it looks up exchange rates at block times where the user sent or received lp tokens, then can calculate a value for those transfers